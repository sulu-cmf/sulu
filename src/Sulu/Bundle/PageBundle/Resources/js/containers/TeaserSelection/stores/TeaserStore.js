// @flow
import {action, autorun, observable} from 'mobx';
import type {IObservableValue} from 'mobx';
import ResourceRequester from 'sulu-admin-bundle/services/ResourceRequester';
import type {TeaserItem} from '../types';

export default class TeaserStore {
    locale: IObservableValue<string>;
    @observable teaserItemIds: Array<{id: number | string, type: string}> = [];
    @observable teaserItems: Array<TeaserItem> = [];
    @observable loading: boolean = false;
    teaserDisposer: () => void;

    constructor(locale: IObservableValue<string>) {
        this.locale = locale;
        this.teaserDisposer = autorun(this.loadTeasers);
    }

    destroy() {
        this.teaserDisposer();
    }

    loadTeasers = () => {
        this.setLoading(true);
        ResourceRequester.getList(
            'teasers',
            {
                ids: this.teaserItemIds.map((teaserItemId) => teaserItemId.type + ';' + teaserItemId.id),
                locale: this.locale.get(),
            }
        ).then(action((response) => {
            this.teaserItems.splice(0, this.teaserItems.length, ...response._embedded.teasers);
            this.setLoading(false);
        }));
    };

    add(type: string, id: number | string) {
        if (this.teaserItemIds.find((teaserItemId) => teaserItemId.type === type && teaserItemId.id === id)) {
            return;
        }

        this.teaserItemIds.push({type, id});
    }

    findById(type: string, id: number | string) {
        return this.teaserItems.find((teaserItem) => teaserItem.type === type && teaserItem.id === id);
    }

    @action setLoading(loading: boolean) {
        this.loading = loading;
    }
}
