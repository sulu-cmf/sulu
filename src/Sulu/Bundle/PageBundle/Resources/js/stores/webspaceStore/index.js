// @flow
import WebspaceStore from './webspaceStore';

export default WebspaceStore;
