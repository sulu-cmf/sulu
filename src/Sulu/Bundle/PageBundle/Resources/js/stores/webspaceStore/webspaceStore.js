// @flow
import {computed, observable} from 'mobx';
import log from 'loglevel';
import type {Webspace} from './types';

class WebspaceStore {
    @observable allWebspaces: Array<Webspace>;

    setWebspaces(webspaces: Array<Webspace>) {
        this.allWebspaces = webspaces;
    }

    @computed get grantedWebspaces(): Array<Webspace> {
        return this.allWebspaces.filter((webspace) => {
            return webspace._permissions.view === true;
        });
    }

    hasWebspace(webspaceKey: string): boolean {
        return !!this.allWebspaces.find((webspace) => webspace.key === webspaceKey);
    }

    getWebspace(webspaceKey: string): Webspace {
        const webspace = this.allWebspaces.find((webspace) => webspace.key === webspaceKey);

        if (!webspace) {
            throw new Error('Webspace "' + webspaceKey + '" not found');
        }

        return webspace;
    }

    // @deprecated
    loadWebspaces(): Promise<Array<Webspace>> {
        log.warn(
            'The "loadWebspaces" method is deprecated since 2.1 and will be removed. ' +
            'Use the "grantedWebspaces" property instead.'
        );

        return Promise.resolve(this.grantedWebspaces);
    }

    // @deprecated
    loadWebspace(webspaceKey: string): Promise<Webspace> {
        log.warn(
            'The "loadWebspace" method is deprecated since 2.1 and will be removed. ' +
            'Use the "getWebspace" method instead.'
        );

        return Promise.resolve(this.getWebspace(webspaceKey));
    }
}

export default new WebspaceStore();
