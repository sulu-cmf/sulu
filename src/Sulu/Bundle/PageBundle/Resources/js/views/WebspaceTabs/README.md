The `WebspaceTabs` view acts as a decorator for the [`Tabs` view](#tabs-1). In addition to rendering the tab itself it
will display a [`WebspaceSelect`](#webspaceselect), based on which a webspace can be chosen. This webspace will then be
passed along with the `webspaceKey` observable to the `children` function, so that the child view can use this
information.
