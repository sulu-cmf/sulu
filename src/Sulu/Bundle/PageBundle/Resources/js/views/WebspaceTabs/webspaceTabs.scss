@import './variables.scss';

.webspace-select {
    position: relative;
    z-index: 2;
    display: inline-block;
    margin-bottom: $webspaceTabsWebspaceSelectMargin;
    width: 0;
}
