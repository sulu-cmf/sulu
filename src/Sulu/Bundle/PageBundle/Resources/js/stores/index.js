// @flow
import webspaceStore from './webspaceStore';

export {webspaceStore};
