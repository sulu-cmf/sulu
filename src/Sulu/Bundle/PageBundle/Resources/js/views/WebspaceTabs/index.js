// @flow
import WebspaceTabs from './WebspaceTabs';

export default WebspaceTabs;
