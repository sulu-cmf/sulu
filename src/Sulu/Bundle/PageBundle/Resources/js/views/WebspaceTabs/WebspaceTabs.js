// @flow
import React from 'react';
import {action, computed, intercept, observable} from 'mobx';
import type {IObservableValue} from 'mobx';
import {observer} from 'mobx-react';
import {Loader} from 'sulu-admin-bundle/components';
import type {ViewProps} from 'sulu-admin-bundle/containers';
import type {AttributeMap, Route} from 'sulu-admin-bundle/services';
import {userStore} from 'sulu-admin-bundle/stores';
import {Tabs} from 'sulu-admin-bundle/views';
import WebspaceSelect from '../../components/WebspaceSelect';
import webspaceStore from '../../stores/webspaceStore';
import type {Webspace} from '../../stores/webspaceStore/types';
import webspaceTabsStyles from './webspaceTabs.scss';

const USER_SETTING_PREFIX = 'sulu_page.webspace_tabs';
const USER_SETTING_WEBSPACE = [USER_SETTING_PREFIX, 'webspace'].join('.');

@observer
class WebspaceTabs extends React.Component<ViewProps> {
    @observable webspaces: ?Array<Webspace>;
    webspaceKey: IObservableValue<string> = observable.box();
    webspaceDisposer: () => void;
    bindWebspaceToRouterDisposer: () => void;

    static getDerivedRouteAttributes(route: Route, attributes: AttributeMap) {
        const webspace = attributes.webspace
            ? attributes.webspace
            : userStore.getPersistentSetting(USER_SETTING_WEBSPACE);

        return {webspace};
    }

    @computed get webspace() {
        if (!this.webspaces) {
            return undefined;
        }

        return this.webspaces.find((webspace) => webspace.key === this.webspaceKey.get());
    }

    componentDidMount() {
        const {router} = this.props;

        this.bindWebspaceToRouter();

        webspaceStore.loadWebspaces()
            .then(action((webspaces) => {
                this.webspaces = webspaces;
            }));

        this.webspaceDisposer = intercept(this.webspaceKey, '', (change) => {
            if (!change.newValue) {
                return change;
            }

            userStore.setPersistentSetting(USER_SETTING_WEBSPACE, change.newValue);
            return change;
        });

        this.bindWebspaceToRouterDisposer = router.addUpdateRouteHook(this.bindWebspaceToRouter);
    }

    componentWillUnmount() {
        this.bindWebspaceToRouterDisposer();
        this.webspaceDisposer();
    }

    bindWebspaceToRouter = () => {
        const {router} = this.props;
        router.bind('webspace', this.webspaceKey);

        return true;
    };

    @action handleWebspaceChange = (value: string) => {
        this.webspaceKey.set(value);
    };

    render() {
        return this.webspaces
            ? (
                <Tabs
                    {...this.props}
                    childrenProps={{webspace: this.webspace, webspaceKey: this.webspaceKey}}
                    header={
                        <div className={webspaceTabsStyles.webspaceSelect}>
                            <WebspaceSelect onChange={this.handleWebspaceChange} value={this.webspaceKey.get()}>
                                {this.webspaces.map((webspace: Webspace) => (
                                    <WebspaceSelect.Item key={webspace.key} value={webspace.key}>
                                        {webspace.name}
                                    </WebspaceSelect.Item>
                                ))}
                            </WebspaceSelect>
                        </div>
                    }
                />
            )
            : <Loader />;
    }
}

export default WebspaceTabs;
