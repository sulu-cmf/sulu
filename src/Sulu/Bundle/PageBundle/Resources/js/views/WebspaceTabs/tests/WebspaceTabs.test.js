// @flow
import React from 'react';
import {mount, render} from 'enzyme';
import {Router} from 'sulu-admin-bundle/services';
import {userStore} from 'sulu-admin-bundle/stores';
import WebspaceTabs from '../WebspaceTabs';
import webspaceStore from '../../../stores/webspaceStore';

jest.mock('sulu-admin-bundle/services/Router', () => jest.fn(function() {
    this.addUpdateRouteHook = jest.fn();
    this.bind = jest.fn();
}));

jest.mock('../../../stores/webspaceStore', () => ({
    loadWebspaces: jest.fn(() => Promise.resolve()),
}));

jest.mock('sulu-admin-bundle/stores/userStore', () => ({
    setPersistentSetting: jest.fn(),
    getPersistentSetting: jest.fn(),
}));

test('Render loader when webspaces are not loaded yet', () => {
    const router = new Router({});

    expect(render(<WebspaceTabs route={router.route} router={router} />)).toMatchSnapshot();
});

test('Render webspace select with children when webspaces are not loaded yet', () => {
    const router = new Router({});

    const route = {
        attributeDefaults: {},
        children: [],
        name: 'webspace_tabs',
        options: {},
        parent: undefined,
        path: '/webspace_tabs',
        rerenderAttributes: [],
        type: 'webspace_tabs',
    };

    const promise = Promise.resolve(
        [
            {key: 'sulu', localizations: [{locale: 'en', default: true}]},
            {key: 'sulu_blog', localizations: [{locale: 'en', default: false}, {locale: 'de', default: true}]},
        ]
    );
    webspaceStore.loadWebspaces.mockReturnValue(promise);

    const webspaceTabs = mount(
        <WebspaceTabs route={route} router={router}>
            {(props) => <h1>{props && props.webspace && props.webspace.key}</h1>}
        </WebspaceTabs>
    );

    webspaceTabs.instance().webspaceKey.set('sulu_blog');

    return promise.then(() => {
        webspaceTabs.update();
        expect(webspaceTabs.children().render()).toMatchSnapshot();
    });
});

test('Load webspace userStore if no route attribute is given', () => {
    userStore.getPersistentSetting.mockImplementation((key) => {
        if (key === 'sulu_page.webspace_tabs.webspace') {
            return 'sulu';
        }
    });

    // $FlowFixMe
    expect(WebspaceTabs.getDerivedRouteAttributes(undefined, {})).toEqual({webspace: 'sulu'});
});

test('Load webspace from route attributes', () => {
    userStore.getPersistentSetting.mockImplementation((key) => {
        if (key === 'sulu_page.webspace_overview.webspace') {
            return 'sulu';
        }
    });

    // $FlowFixMe
    expect(WebspaceTabs.getDerivedRouteAttributes(undefined, {webspace: 'abc'})).toEqual({webspace: 'abc'});
});

test('Should bind and unbind router attributes and updateRouteHook', () => {
    const router = new Router({});

    const route = {
        attributeDefaults: {},
        children: [],
        name: 'webspace_tabs',
        options: {},
        parent: undefined,
        path: '/webspace_tabs',
        rerenderAttributes: [],
        type: 'webspace_tabs',
    };

    const bindWebspaceToRouterDisposerSpy = jest.fn();
    router.addUpdateRouteHook.mockImplementationOnce(() => bindWebspaceToRouterDisposerSpy);
    const webspaceTabs = mount(<WebspaceTabs route={route} router={router}>{() => null}</WebspaceTabs>);

    expect(router.bind).toBeCalledWith('webspace', webspaceTabs.instance().webspaceKey);
    expect(router.addUpdateRouteHook).toBeCalledWith(webspaceTabs.instance().bindWebspaceToRouter);

    const webspaceDisposer = jest.fn();

    webspaceTabs.instance().webspaceDisposer = webspaceDisposer;

    webspaceTabs.unmount();
    expect(bindWebspaceToRouterDisposerSpy).toBeCalledWith();
    expect(webspaceDisposer).toBeCalledWith();
});

test('Save and update webspace when select value is changed', () => {
    const router = new Router({});

    const route = {
        attributeDefaults: {},
        children: [],
        name: 'webspace_tabs',
        options: {},
        parent: undefined,
        path: '/webspace_tabs',
        rerenderAttributes: [],
        type: 'webspace_tabs',
    };

    const webspaces = [
        {key: 'sulu', localizations: [{locale: 'en', default: true}]},
        {key: 'sulu_blog', localizations: [{locale: 'en', default: false}, {locale: 'de', default: true}]},
    ];

    const promise = Promise.resolve(webspaces);
    webspaceStore.loadWebspaces.mockReturnValue(promise);

    const webspaceTabs = mount(<WebspaceTabs route={route} router={router}>{() => null}</WebspaceTabs>);
    webspaceTabs.instance().webspaceKey.set('sulu_blog');

    return promise.then(() => {
        webspaceTabs.update();
        expect(webspaceTabs.find('WebspaceSelect').prop('value')).toEqual('sulu_blog');
        expect(webspaceTabs.find('Tabs').at(0).prop('childrenProps'))
            .toEqual(expect.objectContaining({webspace: webspaces[1]}));
        webspaceTabs.find('WebspaceSelect').prop('onChange')('sulu');

        webspaceTabs.update();
        expect(userStore.setPersistentSetting).toBeCalledWith('sulu_page.webspace_tabs.webspace', 'sulu');
        expect(webspaceTabs.find('Tabs').at(0).prop('childrenProps'))
            .toEqual(expect.objectContaining({webspace: webspaces[0]}));
        expect(webspaceTabs.find('WebspaceSelect').prop('value')).toEqual('sulu');
    });
});
