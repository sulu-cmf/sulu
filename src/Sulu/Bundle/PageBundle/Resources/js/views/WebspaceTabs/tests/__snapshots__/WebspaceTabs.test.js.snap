// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render loader when webspaces are not loaded yet 1`] = `
<div
  class="spinner"
  style="width:40px;height:40px"
>
  <div
    class="doubleBounce1"
  />
  <div
    class="doubleBounce2"
  />
</div>
`;

exports[`Render webspace select with children when webspaces are not loaded yet 1`] = `
Array [
  <div
    class="tabsContainer"
  >
    <ul
      class="tabs"
    />
  </div>,
  <div
    class="webspaceSelect"
  >
    <div
      class="webspaceSelect"
    >
      <button
        class="button"
      >
        <span
          aria-label="su-webspace"
          class="su-webspace buttonIcon"
        />
        <span
          class="buttonValue"
        />
        <span
          aria-label="su-angle-down"
          class="su-angle-down buttonIcon"
        />
      </button>
    </div>
  </div>,
  <h1>
    sulu_blog
  </h1>,
]
`;
